novawallet/Common/QRCreation/QRCodeFactory.swift

```diff
@@ -0,0 +1,203 @@
+            [(operation: operation, withLogo: true)],
+            completion: completion
+        )
+    }
+
+    func createQRCode(
+        using cache: ImageCache,
+        cacheKey: String,
+        payload: Data,
+        logoInfo: QRLogoInfo,
+        qrSize: CGSize,
+        completion: @escaping (Result) -> Void
+    ) {
+        cache.retrieveImage(forKey: cacheKey) { [weak self] result in
+            guard let self else { return }
+
+            var updatedLogoInfo: QRLogoInfo? = logoInfo
+
+            if
+                case let .success(cacheResult) = result,
+                let image = cacheResult.image,
+                let type = logoInfo.type {
+                let localType: QRLogoType? = switch type {
+                case let .remoteColored(url): .localColored(image)
+                case let .remoteTransparent(url): .localTransparent(image)
+                default: nil
+                }
+
+                updatedLogoInfo = QRLogoInfo(
+                    size: logoInfo.size,
+                    type: localType
+                )
+            }
+
+            let operation = operationFactory.createOperation(
+                payload: payload,
+                logoInfo: updatedLogoInfo,
+                qrSize: qrSize
+            )
+
+            execute(
+                [(operation: operation, withLogo: true)],
+                completion: completion
+            )
+        }
+    }
+
+    func createQRCodes(
+        using payload: Data,
+        logoInfo: QRLogoInfo,
+        qrSize: CGSize,
+        completion: @escaping (Result) -> Void
+    ) {
+        let noLogoQROperation = operationFactory.createOperation(
+            payload: payload,
+            logoInfo: QRLogoInfo(size: logoInfo.size, type: nil),
+            qrSize: qrSize
+        )
+
+        let embeddedLogoQROperation = operationFactory.createOperation(
+            payload: payload,
+            logoInfo: logoInfo,
+            qrSize: qrSize
+        )
+
+        let operations = [
+            (operation: noLogoQROperation, withLogo: false),
+            (operation: embeddedLogoQROperation, withLogo: true)
+        ]
+
+        execute(operations, completion: completion)
+    }
+
+    func execute(
+        _ operations: [(operation: QRCreationOperation, withLogo: Bool)],
+        completion: @escaping (Result) -> Void
+    ) {
+        operations.forEach { operation in
+            novawallet.execute(
+                operation: operation.operation,
+                inOperationQueue: operationQueue,
+                runningCallbackIn: .main
+            ) { result in
+                switch result {
+                case let .success(qrCodeImage):
+                    let qrCreationResult: Result = operation.withLogo
+                        ? .full(qrCodeImage)
+                        : .noLogo(qrCodeImage)
+
+                    completion(qrCreationResult)
+                case let .failure(error):
+                    print(error)
```

Comment:
I think we also need to call completion with an error and log it in presenter if it is unexpected